st3/mdpopups/__init__.py

```diff
@@ -815,3 +819,117 @@ def format_frontmatter(values):
     """Format values as frontmatter."""
 
     return frontmatter.dump_frontmatter(values)
+
+
+class _ImagesToResolveParser(html.parser.HTMLParser):
+
+    def __init__(self):
+        super().__init__()
+        self.images_to_resolve = {}
+
+    def handle_starttag(self, tag, attrs):
+        if tag == "img":
+            for name, value in attrs:
+                if name == "src":
+                    if urllib.parse.urlparse(value).scheme in ("http", "https"):
+                        self.images_to_resolve.setdefault(value, []).append(self.getpos())
+                        return
+
+
+class _ReplaceUrlParser(html.parser.HTMLParser):
+
+    def __init__(self, images):
+        super().__init__()
+        self.images = images
+        self.result = ""
+
+    def _render_tag(self, tag, attrs, startend):
+        if tag == "img":
+            for index, (name, value) in enumerate(attrs):
+                if name == "src":
+                    if urllib.parse.urlparse(value).scheme in ("http", "https"):
+                        embbeded = "data:{0}, {1}".format(*self.images[value])
+                        attrs[index] = (name, embbeded)
+        if attrs:
+            rendered = " ".join('{}="{}"'.format(k, v) for k, v in attrs)
+            self.result += "<" + tag + " " + rendered + startend + ">"
+        else:
+            self.result += "<" + tag + startend + ">"
+
+    def handle_starttag(self, tag, attrs):
+        self._render_tag(tag, attrs, "")
+
+    def handle_startendtag(self, tag, attrs):
+        self._render_tag(tag, attrs, "/")
+
+    def handle_endtag(self, tag):
+        self.result += "</" + tag + ">"
+
+    def handle_data(self, data):
+        self.result += data
+
+
+class _ImageResolver:
+
+    def __init__(self, minihtml, resolver, done_callback, images_to_resolve):
+        self.minihtml = minihtml
+        self.done_callback = done_callback
+        self.images_to_resolve = images_to_resolve
+        self.resolved_images = {}
+        for url in self.images_to_resolve.keys():
+            resolver(url, functools.partial(self.on_image_resolved, url))
+
+    def on_image_resolved(self, url, data):
+        path = urllib.parse.urlparse(url).path.lower()
+        if path.endswith(".jpg") or path.endswith(".jpeg"):
+            mime = "image/jpeg"
+        elif path.endswith(".png"):
+            mime = "image/png"
+        elif path.endswith(".gif"):
+            mime = "image/gif"
+        else:
+            # last effort
+            mime = "image/png"
```

Comment:
The mime type should probably be handled by the resolver function (by reading the Content-Type header of the HTTP response).